### DISCUSSION TITLE
Port to Rust for Faster Performance

### DISCUSSION THREAD

**Discussion by sno2** [windicss/windicss#520] [Ideas]
Hello, I have been using Rust for 'bout 2 years now and love it.  Not only is it a fantastic language, but it has a great compiler, fluid language design, and awesome performance (native + wasm).  Anyways, I started working on a prototype port of Tailwind CSS JIT engine using Rust about a month ago.  School was pretty busy the last couple of weeks and I was forced to set it aside.  However, I just found this framework and I think that it might be cooler if I instead modified the Rust code to reproduce windicss's behavior as JIT is first-class in windicss in contrast to Tailwind.  Anyways, I would be willing to modify my code to work for windicss as I built a plugin (dynamic + static classes) system and mixin (prefixes i.e. `dark:` or `md:`) system.  Here is a small example of how the system works ATM:

```rust
use windicss::*;

// `DisplayPlugin` source

define_static_plugin! { // a custom macro I created
  DisplayPlugin {
    "block" { // defines `block` class
      "display": "block"
    }
    "inline-block" {
      "display": "inline-block"
    }
  }
}

// similar to `DarkMixin` source

pub struct DarkMixin;

impl Mixin for DarkMixin {
  const MEDIA: &str = "prefers-color-scheme:dark";

  fn try_generate_media(&self, mixin_namespace: &str) -> Result<&str, ()> { // this media would be included
    match mixin_namespace {
      "dark" => Ok(MEDIA),
      _ => Err(())
    }
  }
}

fn main() {
  let plugins: &[&dyn Plugin] = &[&DisplayPlugin]; // our slice of `Plugin` implementors
  let mixins: &[&dyn Mixin] = &[&DarkMixin]; // our slice of `Mixin` implementors
  let compiler = Compiler::new(plugins, mixins);
  println!("{:?}", compiler.generate(&["block"])); // takes in a slice of class names
  // > ".block{display:block}"
  println!("{:?}", compiler.generate(&["dark:block"])); // this would even allow composable mixins (i.e. dark:md:block) and just include all medias
  // > "@media(prefers-color-scheme:dark){.dark\:block{display:block}}"
}
```

Basically, we would loop through all plugins/mixins and try to generate either a style or media for them.  Thanks for reading and I hope you consider my proposal!

Edit: Currently I have not worked on anything related to using this in JavaScript (web/node); however, I think that creating a plugin in JS and then using in Rust would be easier as plugins/mixins are just raw structs so they can store JS values/functions.

_Originally posted in #519 but I figured GitHub discussions was a better home._

**Comment by antfu** [comment]
Thanks for bringing this up. I know that the front-end community is moving to native build tools for performance (esbuild, swc, etc.) - they are indeed great on performance but on the other hand, the plugin system is something that it's not completely solved. Meanwhile, the performance gain of moving to native might not apply to all the tasks. For computationally intensive tasks like parsing and transforming it might be worth it, but I'd actually doubt that on atomic-css generation. Coincidently, I am also working on a new engine for Windi but purely in JS: https://twitter.com/antfu7/status/1445368569550688256 (which will be public soon) - I think this is kinda proving that the algorithm and architecture might sometimes be the actual bottleneck of performance. So as a conclusion, I'd like to see tools written in Rust or Go for better performance but I don't think at the current stage Windi would need to trade the flexibility and customization for the performance on language - there is still a lot of room for improvements in the JS world.

And finally, I don't want to discourage your enthusiasm. Please feel free to continue your experiments and maybe you could prove me wrong. Or if you want, I am happy to have you collaborating on UnoCSS and see what we could got in the balance of perf and flexibility together. Thanks

**Reply by sno2** [reply]
Thanks for the response and your insights.  I just saw your post on Twitter about unocss and checked to see the discussion and your reply.  I agree with you on the fact that WindiCSS would probably not be a good fit with it's current API and am going to look into how unocss and explore how it may be implemented in Rust.  Thanks
